records/2025-04-14.batch.saved.ses.receipts.md

```diff
@@ -0,0 +1,80 @@
+
+To enable our callback lambda to receive batches of receipts, a new SQS queue is introduced sitting between the SNS topic and lambda, and subscriptions are re-arranged:
+
+1. The lambda subscribes to the new `SQS buffer queue` instead of the SNS topic
+2. The `SQS buffer queue` subscribes to the SNS topic
+
+![current implementation](./diagrams/2025-04-14.batch.saved.ses.receipts/proposed-added-sqs-queue.png)
+
+To achieve the desired batching behaviour, the lambda's subscription to the new SQS buffer queue is configured with:
+- A [BATCH_SIZE](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_event_source_mapping#batch_size-1) indicating how many messages to transmit when calling the lambda
+- A [`MAX_BATCHING_WINDOW`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_event_source_mapping#maximum_batching_window_in_seconds-1) indicating how long to wait between calls to the lambda
+
+The SQS queue continually receives receipt events from the SNS topic and flushes them to the lambda based on the conditions defined by the subscription config:
+- The # of messages in queue is >= `BATCH_SIZE`
+- The time since the last call to the lambda is >= to the `MAX_BATCHING_WINDOW`
+    - Allows us to call the lambda even when the # of messages in the SQS queue is less than the configured `BATCH_SIZE` preventing unnecessary delays.
+
+
+### Option 2 - Leverage Redis as our buffer queue
+Instead of adding a new SQS queue to buffer between the SNS topic and Lambda, Redis is leveraged as the "buffer queue" instead.
+
+1. The lambda remains connected to the SNS topic, and receives receipts 1 by 1 like before
+2. Instead of immediately packaging the receipt as a celery task, the lambda adds the receipt to Redis.
+3. When the # of receipts in Redis >= to the `BATCH_SIZE` the receipts are flushed from Redis, and sent to `process_ses_results` as a celery task similar to the existing implementation
+
+## Additional considerations
+
+**The added SQS queue approach** makes infrastructure changes to a critical part of the system with the sole purpose being to enable the lambda to receive receipts in batches. However once set up, only minor adjustments to the current implementation are needed to be up and running.
+
+**The latter approach** leaves infrastructure mostly untouched but presents a number of new implementation challenges in enabling Redis functionality in the lambdas:
+1. Currently our callback lambdas do not have access to the utils project, which is where our Redis client code lives. Depending on implementation details, this could result in duplicated Redis client code and diverges from our current architecture of keeping the lambdas lightweight and with a specialized purpose.
+2. Redis connection secrets need to be added to the lambda's environment variables
+3. VPC configuration would need to change to give the lambdas to access the Redis cluster
```

Comment:
Having read through both these approaches and the problem you are trying to solve based on your testing for batch saving, I think option 1 will be the easier, option. It also builds on how the lambda currently uses this infrastructure. I see no significant performance advantages of option 1 over option 2. My suggestion is to go with option 1.

Reply:
Option 1 seems better to me as well, considering this is faster delivery, less infrastructure setup compared to option 2, while upgrading the receipts' scaling properties.

Reply:
Moving forward with Option 1. I've added it to the decision section.